include/seqan3/alignment/scoring/detail/simd_matrix_scoring_scheme.hpp

```diff
@@ -77,115 +114,107 @@ class simd_matrix_scoring_scheme
+        [[maybe_unused]] auto check_score_range = [&] ([[maybe_unused]] score_t score)
         {
-            if (min_or_max_exceeded<score_t, simd_scalar_t>(scoring_scheme))
-                throw std::invalid_argument{"The selected scoring scheme score overflows "
-                                            "for the selected scalar type of the simd type."};
-        }
+            // Note only if the size of the scalar type of the simd vector is smaller than the size of the score type
+            // of the original scoring scheme, the score might exceed the valid value range of the scalar type. In this
+            // case an exception will be thrown.
+            if constexpr (sizeof(scalar_type) < sizeof(score_t))
+            {
+                constexpr score_t max_score_value = static_cast<score_t>(std::numeric_limits<scalar_type>::max());
+                constexpr score_t min_score_value = static_cast<score_t>(std::numeric_limits<scalar_type>::lowest());
 
-        internal_scoring_scheme = scoring_scheme;
-    }
+                if (score > max_score_value || score < min_score_value)
+                    throw std::invalid_argument{"The selected scoring scheme score overflows "
+                                                "for the selected scalar type of the simd type."};
+            }
+        };
 
-    /*!\brief Check if any score in the scoring scheme matrix exceeds the min or max value allowed by the simd vector.
-     * \param[in] scoring_scheme The scoring scheme to check for min or max violations.
-     * \tparam score_t The type of the scores in the scoring scheme matrix.
-     * \tparam simd_scalar_t The type of the simd vector.
-     * \returns Returns `true` if it encounters a value larger than the maximum or lower than the minimum allowed by
-     *          the simd vector, and `false` if otherwise.
-     */
-    template <typename score_t, typename simd_scalar_t>
-    constexpr bool min_or_max_exceeded(scoring_scheme_t const & scoring_scheme)
-    {
-        score_t max_val = static_cast<score_t>(std::numeric_limits<simd_scalar_t>::max());
-        score_t min_val = static_cast<score_t>(std::numeric_limits<simd_scalar_t>::lowest());
+        // For the global alignment we extend the alphabet by one symbol to handle sequences with different size.
+        scoring_scheme_data.resize(index_offset * index_offset, score_for_padding_symbol);
 
-        for (uint8_t i = 0; i < alphabet_size<alphabet_t>; ++i)
+        // Convert the scoring matrix into a linear vector to allow gather operations later on.
+        auto data_it = scoring_scheme_data.begin();
+        for (rank_type lhs_rank = 0; lhs_rank < seqan3::alphabet_size<alphabet_t>; ++lhs_rank)
         {
-            for (uint8_t j = 0; j < alphabet_size<alphabet_t>; ++j)
+            for (rank_type rhs_rank = 0; rhs_rank < seqan3::alphabet_size<alphabet_t>; ++rhs_rank, ++data_it)
```

Comment:
I just remember in the back of my head that we might have had issues with this.. 
So the rank_type is e.g. `uint8_t`, which can represent `256` ranks from `0 - 255`. If I now have an alphabet of size `256` (which still fits), is the comparison `rhs_rank < seqan3::alphabet_size<alphabet_t>` still safe? Since the right side might be cast into `rank_type` and overflow. I'm not familiar with the exact cast when comparing integers, maybe you give it a second thought just to be sure?

Reply:
this should be fine. In any case the sizeof the alphabet_size type will be at least as big as the size of rank_type. In this case the comparison is anyway promoted to the higher integer type but at least to a 32 bit on most systems if I recall correctly. Otherwise the implementation of the semialphabet is broken, but then the concept is broken as well even if the semantic requirements cannot be explicitly checked. Only issue that might raise, though, is that no one requires both to have the same signedness.